src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/Delegation/TextEditResponseRewriter.cs

```diff
@@ -0,0 +1,97 @@
+
+        public override async Task<VSInternalCompletionList> RewriteAsync(
+            VSInternalCompletionList completionList,
+            int hostDocumentIndex,
+            DocumentContext hostDocumentContext,
+            DelegatedCompletionParams delegatedParameters,
+            CancellationToken cancellationToken)
+        {
+            if (delegatedParameters.ProjectedKind != RazorLanguageKind.CSharp)
+            {
+                return completionList;
+            }
+
+            var sourceText = await hostDocumentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+
+            sourceText.GetLineAndOffset(hostDocumentIndex, out var lineNumber, out var characterOffset);
+            var hostDocumentPosition = new Position(lineNumber, characterOffset);
+            completionList = TranslateTextEdits(hostDocumentPosition, delegatedParameters.ProjectedPosition, completionList);
+
+            if (completionList.ItemDefaults?.EditRange != null)
+            {
+                completionList.ItemDefaults.EditRange = TranslateRange(hostDocumentPosition, delegatedParameters.ProjectedPosition, completionList.ItemDefaults.EditRange);
+            }
+
+            return completionList;
+        }
+
+        // The TextEdit positions returned to us from the C#/HTML language servers are positions correlating to the virtual document.
+        // We need to translate these positions to apply to the Razor document instead. Performance is a big concern here, so we want to
+        // make the logic as simple as possible, i.e. no asynchronous calls.
+        // The current logic takes the approach of assuming the original request's position (Razor doc) correlates directly to the positions
+        // returned by the C#/HTML language servers. We use this assumption (+ math) to map from the virtual (projected) doc positions ->
+        // Razor doc positions.
+        internal static VSInternalCompletionList TranslateTextEdits(
+            Position hostDocumentPosition,
+            Position projectedPosition,
+            VSInternalCompletionList completionList)
+        {
+            foreach (var item in completionList.Items)
+            {
+                if (item.TextEdit != null)
+                {
+                    var translatedRange = TranslateRange(hostDocumentPosition, projectedPosition, item.TextEdit.Range);
+                    item.TextEdit.Range = translatedRange;
+                }
+                else if (item.AdditionalTextEdits != null)
+                {
+                    // Additional text edits should typically only be provided at resolve time. We don't support them in the normal completion flow.
+                    item.AdditionalTextEdits = null;
+                }
+            }
+
+            return completionList;
+        }
+
+        internal static Range TranslateRange(Position hostDocumentPosition, Position projectedPosition, Range textEditRange)
```

Comment:
Why internal? Doesn't seem to be used from tests, so should be private. If it's used from another type, it should be public.